peer/voting/voting.go

```diff
@@ -0,0 +1,288 @@
+}
+
+// NumVoteOfDigest - number of votest for a digest
+func (v *VotingImpl) NumVoteOfDigest(digest blockdigest.Digest) int {
+	if v.existVoteForDigest(digest) {
+		return len(v.votes[digest])
+	}
+	return 0
+}
+
+func (v *VotingImpl) existVoteForDigest(digest blockdigest.Digest) bool {
+	_, ok := v.votes[digest]
+	return ok
+}
+
+// VoteBy - vote by some upstream
+func (v *VotingImpl) VoteBy(candidate upstream.UpstreamIntf) {
+	height := candidate.CachedRemoteHeight()
+	digest := candidate.CachedRemoteDigestOfLocalHeight()
+	remoteAddr := candidate.Client().String()
+	upstream := candidate.Name()
+	v.log.Infof(
+		"%s connects to remote %s, cached remote height: %d with digest: %x",
+		upstream,
+		remoteAddr,
+		height,
+		digest,
+	)
+
+	if !v.validHeight(height) {
+		v.log.Warnf(
+			"remote cached height: %d, below minimum height %d, discard",
+			height,
+			v.minHeight,
+		)
+		return
+	}
+
+	e := &voters{
+		candidate: candidate,
+		height:    height,
+	}
+
+	if v.existVoteForDigest(digest) {
+		v.votes[digest] = append(v.votes[digest], e)
+		return
+	}
+
+	v.log.Debugf("%s connect to remote %s, vote success", upstream, remoteAddr)
+	v.votes[digest] = []*voters{e}
+}
+
+func (v *VotingImpl) validHeight(height uint64) bool {
+	return height >= v.minHeight
+}
+
+// ElectedCandidate - get candidate that is most vote
+func (v *VotingImpl) ElectedCandidate() (upstream.UpstreamIntf, uint64, error) {
+	err := v.countVotes()
+	if nil != err {
+		v.log.Errorf("count votes with error: %s", err)
+		return nil, uint64(0), err
+	}
+
+	if v.result.draw {
+		return v.drawElection()
+	}
+
+	return v.result.winner, v.result.winner.CachedRemoteHeight(), nil
+}
+
+func (v *VotingImpl) countVotes() error {
+	for _, voters := range v.votes {
+		if v.result.highestNumVotes < len(voters) {
+			v.updateTemporarilyVoteSummary(voters)
+		} else if v.result.highestNumVotes == len(voters) {
+			v.result.draw = true
+		}
+	}
+	v.log.Debugf(
+		"vote draw: %t, most votes: %d, majority height: %d",
+		v.result.draw,
+		v.result.highestNumVotes,
+		v.result.majorityHeight,
+	)
+
+	if nil == v.result.winner {
+		return fmt.Errorf("%s", fault.ErrVotesEmptyWinner)
+	}
+
+	if !v.sufficientVotes() {
+		return fmt.Errorf("%s", fault.ErrVotesInsufficient)
+	}
+
+	return nil
+}
+
+func (v *VotingImpl) sufficientVotes() bool {
+	fmt.Printf("highest num votes: %d\n", v.result.highestNumVotes)
+	fmt.Printf("half min clients: %f\n", math.Ceil(minimumClients/2))
+	fmt.Printf("highest < half min: %t\n", float64(v.result.highestNumVotes) < math.Ceil(minimumClients/2))
+	if !v.result.draw && float64(v.result.highestNumVotes) < math.Ceil(minimumClients/2) {
+		return false
+	}
+
+	return v.sufficientVotesInDraw()
+}
+
+func (v *VotingImpl) sufficientVotesInDraw() bool {
+	drawVotes := 0
+	for _, voters := range v.votes {
+		counts := len(voters)
+		if v.result.highestNumVotes == counts {
+			drawVotes += counts
+		}
+	}
+	if float64(drawVotes) >= math.Ceil(minimumClients/2) {
```

Comment:
Can we simplify the returning boolean expression to pass the static analyser??
Instead of:
`if float64(drawVotes) >= math.Ceil(minimumClients/2) {
        return true
}
return false
` 

Simplify as:
`return float64(drawVotes) >= math.Ceil(minimumClients/2)
`

Reply:
changed